[docs] Add deckhouse module architecture section (#20748)
deckhouse/deckhouse: docs/documentation/pages/architecture/deckhouse/DECKHOUSE.md

---
title: Deckhouse module
permalink: en/architecture/deckhouse/deckhouse.html
search: deckhouse, deckhouse-controller, modules
description: Architecture of the deckhouse module in Deckhouse Kubernetes Platform.
---

The [`deckhouse`](/modules/deckhouse/) module implements the core of Deckhouse Kubernetes Platform (DKP), performing the following operations:
- Platform updates.
- Module configuration management.
- Module installation and updates.
- Module documentation build triggering.

> - Validation of custom resources managed by DKP modules.

Reviewer: ```suggestion
The [`deckhouse`](/modules/deckhouse/) module implements the core of Deckhouse Kubernetes Platform (DKP), performing the following operations:

- Platform updates
- Module configuration management
- Module installation and updates
- Module documentation build triggering
- Validation of custom resources managed by DKP modules
```

The module manages the following custom resources in the `deckhouse.io` API group:

- Module management:
  - [Module](../../reference/api/cr.html#module): Description, status, and publication of module information.
  - [ModuleConfig](../../reference/api/cr.html#moduleconfig): Description of user-defined module settings.
  - [ModulePullOverride](../../reference/api/cr.html#modulepulloverride): Description of module version selection overrides.
  - [ModuleRelease](../../reference/api/cr.html#modulerelease): Description, publication, and tracking of module releases.
  - [ModuleSettingsDefinition](../../reference/api/cr.html#modulesettingsdefinition): Schema, versions, and transformation rules for module settings.
  - [ModuleSource](../../reference/api/cr.html#modulesource): Description of a module source, repository, or storage.
  - [ModuleUpdatePolicy](../../reference/api/cr.html#moduleupdatepolicy): Rules for module updates and version transition automation.

- Platform management:

>   - [DeckhouseRelease](../../reference/api/cr.html#deckhouserelease): An object that defines the Deckhouse release (version) and platform update policy.

Reviewer: ```suggestion
  - [DeckhouseRelease](../../reference/api/cr.html#deckhouserelease): An object that defines the DKP release (version) and platform update policy.
```

  - [ValidationWebhook](/modules/deckhouse/latest/cr.html#validationwebhook): Settings and handlers for resource validation webhooks.

## Module architecture

{% alert level="info" %}
The following simplifications are made in the diagram:

* The diagram shows containers in different pods interacting directly with each other. In reality, they communicate via the corresponding Kubernetes Services (internal load balancers). Service names are omitted if they are obvious from the context. Otherwise, the Service name is shown above the arrow.
* Pods may run multiple replicas. However, each pod is shown as a single replica in the diagram.
{% endalert %}

The Level 2 C4 architecture of the [`deckhouse`](/modules/deckhouse/) module and its interaction with other DKP components are shown in the following diagram:

![Deckhouse module architecture](../../images/architecture/deckhouse/c4-l2-deckhouse-deckhouse.svg)

## Module components

The module consists of the following components:

1. **Deckhouse** (Deployment): A controller that implements platform management operations.

>    The controller orchestrates platform management tasks using [the queueing mechanism](./queues.html).

Reviewer: ```suggestion
   The controller coordinates platform management tasks using the [queueing mechanism](./queues.html).
```

   If [High Availability (HA)](../../admin/configuration/high-reliability-and-availability/) mode is enabled, multiple Deckhouse controller instances are started. To ensure correct behavior, Deckhouse controllers perform leader election using the `deckhouse-leader-election` Lease resource. The controller elected as leader performs all platform management operations.

   In addition, the Deckhouse controller configures:

   | Description       | Module configuration parameter                |
   |-------------- |-------------------------------------- |
   | Logging level          | [`.spec.settings.logLevel`](/modules/deckhouse/configuration.html#parameters-loglevel)   |
   | Default enabled module set | [`.spec.settings.bundle`](/modules/deckhouse/configuration.html#parameters-bundle)   |
   | Release channel | [`.spec.settings.releaseChannel`](/modules/deckhouse/configuration.html#parameters-releasechannel)   |
   | Update mode | [`.spec.settings.update.mode`](/modules/deckhouse/configuration.html#parameters-update-mode)   |

>    | Update windows | [`.spec.settings.update.windows.days`](/modules/deckhouse/configuration.html#parameters-update-windows)   |

Reviewer: ```suggestion
   In addition, you can set the following parameters using the Deckhouse controller:

   | Module configuration parameter | Description |
   | ------------------------------ | ----------- |
   | [`logLevel`](/modules/deckhouse/configuration.html#parameters-loglevel) | Logging level |
   | [`bundle`](/modules/deckhouse/configuration.html#parameters-bundle) | Set of modules enabled by default |
   | [`releaseChannel`](/modules/deckhouse/configuration.html#parameters-releasechannel) | Release channel |
   | [`update.mode`](/modules/deckhouse/configuration.html#parameters-update-mode) | Update mode |
   | [`update.windows.days`](/modules/deckhouse/configuration.html#parameters-update-windows) | Update windows |
```

   For details about module settings, refer to the [module documentation section](/modules/deckhouse/).

   It consists of the following containers:

   * **init-downloaded-modules**: Init container that prepares the directory structure required for module operations.
   * **deckhouse**: Main container.
   * **kube-rbac-proxy**: Sidecar container with an authorization proxy based on Kubernetes RBAC that provides secure access to the main container component debug interface.
  
1. **Webhook-handler** (Deployment): Consists of a single **handler** container and implements a generic webhook for conversion and validation of custom resources managed by DKP.

>     The component watches [ConversionWebhook](/modules/deckhouse/latest/cr.html#conversionwebhook) and [ValidationWebhook](/modules/deckhouse/latest/cr.html#validationwebhook) custom resources and, based on them, generates hook Python files for [shell-operator](https://github.com/flant/shell-operator) from templates. When `kube-apiserver` sends resource validation or conversion requests, shell-operator runs the required hook and returns the processing result.

Reviewer: ```suggestion
   The component watches [ConversionWebhook](/modules/deckhouse/latest/cr.html#conversionwebhook) and [ValidationWebhook](/modules/deckhouse/latest/cr.html#validationwebhook) custom resources and, based on them, generates hook Python files for [shell-operator](https://github.com/flant/shell-operator) from templates. When `kube-apiserver` sends resource validation or conversion requests, shell-operator runs the required hook and returns the processing result.
```

1. **Cni-migration-manager** (Deployment): An optional component running on control plane nodes, consisting of a single **manager** container. The component manages the network plugin (CNI) switching process in the DKP cluster and records the current state in the CNIMigration custom resource. Migration to Flannel, Simple bridge, and Cilium is supported. For details on switching CNI in the cluster, refer to the [corresponding guide](/products/kubernetes-platform/guides/cni-migration.html).

    {% alert level="info" %}
    The component is created by the `detect-cni-migration` global hook when the CNIMigration custom resource exists. The CNIMigration resource is created manually by an administrator or by running the `d8 network cni-migration switch --to-cni <target cni>` command.

>     {% endalert %}

Reviewer: ```suggestion
   {% alert level="info" %}
   The component is created by the `detect-cni-migration` global hook when the CNIMigration custom resource exists. The CNIMigration resource is created manually by an administrator or by running the `d8 network cni-migration switch --to-cni <target cni>` command.
   {% endalert %}
```

1. **Cni-migration-agent** (DaemonSet): An optional component running on all cluster nodes, consisting of a single **agent** container. The component watches the CNIMigration custom resource and manages the CNINodeMigration custom resource that reflects migration state for a specific node.

    {% alert level="info" %}
    The component is created by the `detect-cni-migration` global hook when the CNIMigration custom resource exists. The CNIMigration resource is created manually by an administrator or by running the `d8 network cni-migration switch --to-cni <target cni>` command.

>     {% endalert %}

Reviewer: ```suggestion
   {% alert level="info" %}
   The component is created by the `detect-cni-migration` global hook when the CNIMigration custom resource exists. The CNIMigration resource is created manually by an administrator or by running the `d8 network cni-migration switch --to-cni <target cni>` command.
   {% endalert %}
```

## Module interactions

The module interacts with the following components:

1. **Kube-apiserver**:
   - Working with custom resources in the `deckhouse.io` API group.
   - Watching Pod and DaemonSet resources, as well as restarting Pods when the network plugin is changed.
   - Watching resources described in the ObjectKeeper custom resource.
   - Creating and updating Lease resources.
   - Creating, deleting, modifying, and watching resources described in DKP modules.
   - Authorizing requests.

1. [**Documentation**](/modules/documentation/): Updating documentation when a DKP module is added or updated.

1. **Image registry**: Retrieving module component images along with metadata when the [`registry`](/modules/registry/) module is installed in Unmanaged mode.

> 1. **`registry` module**: Retrieving module component images along with metadata when the [`registry`](/modules/registry/) module is installed in one of the Direct, Proxy, or Local modes.

Reviewer: ```suggestion
1. **Image registry**: Retrieving module component images along with metadata when the [`registry`](/modules/registry/) module is installed in `Unmanaged` mode.

1. **`registry` module**: Retrieving module component images along with metadata when the [`registry`](/modules/registry/) module is installed in one of the following modes: `Direct`, `Proxy`, or `Local`.
```
